UnitySDK/Assets/ML-Agents/Scripts/Agent.cs

```diff
@@ -625,15 +605,37 @@ void SendInfoToBrain()
 
             if (m_Recorder != null && m_Recorder.record && Application.isEditor)
             {
+                // This is a bit of a hack - if we're in inference mode, compressed observations won't be generated
+                // But we need these to be generated for the recorder. So generate them here.
+                if (m_Info.compressedObservations.Count == 0)
+                {
+                    GenerateSensorData();
+                }
+
                 m_Recorder.WriteExperience(m_Info);
             }
 
             m_Info.textObservation = "";
         }
 
-        public void ClearVisualObservations()
+        /// <summary>
+        /// Generate data for each sensor and store it on the Agent's AgentInfo
+        /// </summary>
+        public void GenerateSensorData()
```

Comment:
Is there a reason the Compressed data is in the AgentInfo ? Would it be easier to generate it in the Communicator and the recorder? Maybe this is good for now.

Reply:
Can you document in the code what the future design should look like ?

Reply:
Maybe we should talk about it some more. I think the Brain (and maybe the presence of the recorder) should determine whether we're in compressed, uncompressed, or write-to-tensor mode; this is something that could be passed as a parameter to CollectObservations, which could call GenerateSensorData.

Reply:
> Would it be easier to generate it in the Communicator and the recorder? 

The recorder operates on AgentInfos, not Agents. 

We could pass the ISensors instead of the CompressedObservations in AgentInfo, and let the AgentInfo consumer decide what to do with them...

Reply:
Added some comments